src/PhotoNim/shapes.nim

```diff
@@ -5,142 +5,196 @@ import std/options
 import camera, geometry
 
 
-type 
+type HitRecord* = object
+    ray*: Ray
+    t_hit*: float32
+    world_pt*: Point3D
+    surface_pt*: Point2D
+    normal*: Normal
+
+proc newHitRecord*(ray: Ray, t: float32, hit_point: Point3D, uv: Point2D, normal: Normal): HitRecord {.inline.} =
+    HitRecord(ray: ray, t_hit: t, world_pt: hit_point, surface_pt: uv, normal: normal)
+
+proc areClose*(a, b: HitRecord): bool {.inline.} = 
+    areClose(a.ray, b.ray) and areClose(a.t_hit, b.t_hit) and 
+    areClose(a.world_pt, b.world_pt) and areClose(a.surface_pt, b.surface_pt) and 
+    areClose(a.normal, b.normal) 
+
+
+type
+    AABB* = tuple[min, max: Point3D]
+
     Shape* = object of RootObj
         transf*: Transformation
+        aabb*: Option[AABB] = none(AABB)
```

Comment:
I like this, we don't always want an axis-aligned bounding box: for simple shapes like spheres it's just too much